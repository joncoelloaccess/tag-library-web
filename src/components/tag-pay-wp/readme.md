# tag-pay-wp

















## Demos

















#### Basic button

















```html








<tag-pay-wp></tag-pay-wp>








```

















```html








<tag-pay-wp dayofweek="Monday"></tag-pay-wp>








<tag-pay-wp dayofweek="Tuesday"></tag-pay-wp>








```

















<!-- Auto Generated Below -->


## Properties

#### dayofweek

string


#### value

string


## Attributes

#### dayofweek

string


#### value

string



----------------------------------------------

*Built with [StencilJS](https://stenciljs.com/)*
