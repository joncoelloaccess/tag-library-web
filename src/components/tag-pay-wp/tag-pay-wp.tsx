import { Component, Prop, State } from "@stencil/core";
import {
  payWpFactory,
  PayWp,
  PayWpState,
  PayWpStates
} from "./tag-pay-wp.util";

const buttonColour = new Map()
  .set(PayWpStates.none, "default")
  .set(PayWpStates.halfDay, "glacier")
  .set(PayWpStates.fullDay, "viridian-green");

@Component({
  tag: "tag-pay-wp",
  styleUrl: "tag-pay-wp.scss",
  shadow: true
})
export class TagPayWp {
  @Prop() dayofweek: string;
  @Prop() value: string;
  @State() payWpMachine: PayWp = payWpFactory(PayWpStates.none);
  @State() state: PayWpState = this.payWpMachine.value;
  @State()
  workingPattern: Map<string, any> = new Map()
    .set(
      PayWpStates.none,
      <svg viewBox="0 0 132 28" xmlns="http://www.w3.org/2000/svg">
        <path
          d="M26.34 18A13.14 13.14 0 0 1 .75 13.87a13.13 13.13 0 0 1 25.59-4.12h79.14a13.14 13.14 0 0 1 25.55 4.23A13.13 13.13 0 0 1 105.41 18H26.34z"
          fill="none"
          stroke="#ced4d9"
          stroke-width="1.5"
        />
      </svg>
    )
    .set(
      PayWpStates.halfDay,
      <svg viewBox="0 0 132 28" xmlns="http://www.w3.org/2000/svg">
        <path
          d="M65.89 9.64h39.59a13.14 13.14 0 0 1 25.55 4.23 13.13 13.13 0 0 1-25.62 4.03H65.89V9.64z"
          fill="none"
          stroke="#ced4d9"
          stroke-width="1.5"
        />
        <path
          d="M26.32 18.05A13.14 13.14 0 0 1 .75 13.87a13.13 13.13 0 0 1 25.6-4.07H65.9v8.25H26.32z"
          fill="#67c2c2"
          stroke="#67c2c2"
          stroke-width="1.5"
          stroke-linecap="butt"
          stroke-miterlimit="1.41"
        />
        <circle
          cx="65.89"
          cy="13.88"
          r="13.13"
          fill="#4eb8b5"
          stroke="#409794"
          stroke-width="1.5"
        />
      </svg>
    )
    .set(
      PayWpStates.fullDay,
      <svg viewBox="0 0 132 28" xmlns="http://www.w3.org/2000/svg">
        <path
          d="M26.34 18A13.14 13.14 0 0 1 .75 13.87a13.13 13.13 0 0 1 25.59-4.12H107V18H26.34z"
          fill="#66c2c2"
          stroke="#67c2c2"
          stroke-width="1.5"
        />
        <circle
          cx="117.91"
          cy="13.88"
          r="13.13"
          fill="#4eb8b5"
          stroke="#409794"
          stroke-width="1.5"
          stroke-linecap="round"
          stroke-miterlimit="1.5"
        />
      </svg>
    );

  changeState() {
    this.state = this.payWpMachine.transition().value;
    this.value = this.state.value;
  }

  render() {
    const progressImg = this.workingPattern.get(this.state.value);
    return (
      <div class="paywp">
        <div class="paywp__container">
          <tag-button
            class="paywp__button"
            accent={buttonColour.get(this.state.value)}
            onClick={() => this.changeState()}
          >
            {this.dayofweek}
          </tag-button>
          {progressImg}
        </div>
      </div>
    );
  }
}
