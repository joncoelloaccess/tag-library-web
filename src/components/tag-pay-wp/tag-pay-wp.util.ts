export enum PayWpStates {
  none = "none",
  halfDay = "half-day",
  fullDay = "full-day"
}

export type PayWpState = {
  value: PayWpStates;
};

export interface PayWp {
  transition: () => PayWp;
  value: PayWpState;
}

export type PayWpFactory = (initialState: PayWpStates) => PayWp;

export const payWpFactory: PayWpFactory = initialState => {
  let internalState: PayWpStates = initialState;
  const cycledValue = (val: PayWpStates): PayWpStates => {
    if (val === PayWpStates.none) return PayWpStates.halfDay;
    if (val === PayWpStates.halfDay) return PayWpStates.fullDay;
    if (val === PayWpStates.fullDay) return PayWpStates.none;
  };
  const transition = (): PayWp => {
    internalState = cycledValue(internalState);
    return {
      transition: transition,
      value: {
        value: internalState
      }
    };
  };
  return {
    transition: transition,
    value: {
      value: initialState
    }
  };
};
