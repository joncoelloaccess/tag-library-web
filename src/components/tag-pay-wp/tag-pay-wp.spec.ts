import { render, flush } from "@stencil/core/testing";
import { TagPayWp } from "./tag-pay-wp";
import { payWpFactory, PayWpStates } from "./tag-pay-wp.util";

describe("tag-pay-wp", () => {
  it("should build", () => {
    expect(new TagPayWp()).toBeTruthy();
  });

  describe("rendering", () => {
    let element;
    beforeEach(async () => {
      element = await render({
        components: [TagPayWp],
        html: '<tag-pay-wp dayofweek="Hello"></tag-pay-wp>'
      });
    });
    it("should render correctly", async () => {
      await flush(element);
      expect(element.outerHTML).toMatchSnapshot();
    });
    it("should cycle state from the default state", async () => {
      await flush(element);
      const btn = element.querySelector("tag-button");
      btn.click();
      await flush(element);
      expect(element.outerHTML).toMatchSnapshot();
    });
  });
});

describe("payWorkingPattern", () => {
  it("should exist", () => {
    expect(payWpFactory).not.toBeUndefined();
  });
  it("should return the initial value and false when zero", () => {
    const payWp = payWpFactory(PayWpStates.none);
    const result = payWp.value;
    expect(result.value).toEqual(PayWpStates.none);
  });
  it("should return the intial value and true when not zero", () => {
    const initialValue = PayWpStates.halfDay;
    const payWp = payWpFactory(initialValue);
    const result = payWp.value;
    expect(result.value).toEqual(initialValue);
  });
  it("should return the intial value + 1 and true when transitioned from zero", () => {
    const initialValue = PayWpStates.halfDay;
    const payWp = payWpFactory(initialValue);
    const result = payWp.transition().value;
    expect(result.value).toEqual(PayWpStates.fullDay);
  });
  it("should return the zero and false when transitioned from 2", () => {
    const initialValue = PayWpStates.fullDay;
    const payWp = payWpFactory(initialValue);
    const result = payWp.transition().value;
    expect(result.value).toEqual(PayWpStates.none);
  });
  it("should return the zero and false when transitioned multiple time from 1", () => {
    const initialValue = PayWpStates.halfDay;
    const payWp = payWpFactory(initialValue);
    const result = payWp.transition().transition().value;
    expect(result.value).toEqual(PayWpStates.none);
  });
  it("should return the 1 and true when transitioned multiple time from 2", () => {
    const initialValue = PayWpStates.fullDay;
    const payWp = payWpFactory(initialValue);
    const result = payWp.transition().transition().value;
    expect(result.value).toEqual(PayWpStates.halfDay);
  });
  it("should return the 1 and true when transitioned multiple with a break time from 2", () => {
    const initialValue = PayWpStates.fullDay;
    const payWp = payWpFactory(initialValue);
    payWp.transition();
    const result = payWp.transition().value;
    expect(result.value).toEqual(PayWpStates.halfDay);
  });
});
