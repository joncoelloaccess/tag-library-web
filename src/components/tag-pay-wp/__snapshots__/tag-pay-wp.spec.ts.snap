// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`tag-pay-wp rendering should cycle state from the default state 1`] = `"<tag-pay-wp dayofweek=\\"Hello\\" data-tag-pay-wp-host=\\"\\" class=\\"hydrated\\"><!----><div class=\\"paywp\\" data-tag-pay-wp=\\"\\"><div class=\\"paywp__container\\" data-tag-pay-wp=\\"\\"><tag-button class=\\"paywp__button\\" accent=\\"glacier\\" data-tag-pay-wp=\\"\\">Hello</tag-button><svg viewBox=\\"0 0 132 28\\" xmlns=\\"http://www.w3.org/2000/svg\\" data-tag-pay-wp=\\"\\"><path d=\\"M65.89 9.64h39.59a13.14 13.14 0 0 1 25.55 4.23 13.13 13.13 0 0 1-25.62 4.03H65.89V9.64z\\" fill=\\"none\\" stroke=\\"#ced4d9\\" stroke-width=\\"1.5\\" data-tag-pay-wp=\\"\\"></path><path d=\\"M26.32 18.05A13.14 13.14 0 0 1 .75 13.87a13.13 13.13 0 0 1 25.6-4.07H65.9v8.25H26.32z\\" fill=\\"#67c2c2\\" stroke=\\"#67c2c2\\" stroke-width=\\"1.5\\" stroke-linecap=\\"butt\\" stroke-miterlimit=\\"1.41\\" data-tag-pay-wp=\\"\\"></path><circle cx=\\"65.89\\" cy=\\"13.88\\" r=\\"13.13\\" fill=\\"#4eb8b5\\" stroke=\\"#409794\\" stroke-width=\\"1.5\\" data-tag-pay-wp=\\"\\"></circle></svg></div></div></tag-pay-wp>"`;

exports[`tag-pay-wp rendering should render correctly 1`] = `"<tag-pay-wp dayofweek=\\"Hello\\" data-tag-pay-wp-host=\\"\\" class=\\"hydrated\\"><!----><div class=\\"paywp\\" data-tag-pay-wp=\\"\\"><div class=\\"paywp__container\\" data-tag-pay-wp=\\"\\"><tag-button class=\\"paywp__button\\" accent=\\"default\\" data-tag-pay-wp=\\"\\">Hello</tag-button><svg viewBox=\\"0 0 132 28\\" xmlns=\\"http://www.w3.org/2000/svg\\" data-tag-pay-wp=\\"\\"><path d=\\"M26.34 18A13.14 13.14 0 0 1 .75 13.87a13.13 13.13 0 0 1 25.59-4.12h79.14a13.14 13.14 0 0 1 25.55 4.23A13.13 13.13 0 0 1 105.41 18H26.34z\\" fill=\\"none\\" stroke=\\"#ced4d9\\" stroke-width=\\"1.5\\" data-tag-pay-wp=\\"\\"></path></svg></div></div></tag-pay-wp>"`;
